### ISSUE TITLE
Can not select namespace in central dashboard - CentralDashboard can't talk to kfam; Use DNS and not environment variables

### ISSUE THREAD

**Issue by Jeffwan** [kubeflow/kubeflow#3900]
/kind bug

**What steps did you take and what happened:**
namespace doesn't have a list even I create profile. namespaces were created in backend. 
Checking central dashboard logs and notice we have this problem. `aws` platform is not support here which makes entire dashboard unavailable. 

```
k logs -f centraldashboard-7bd77d6dd4-trhhg

> kubeflow-centraldashboard@0.0.2 start /app
> npm run serve


> kubeflow-centraldashboard@0.0.2 serve /app
> node dist/server.js

Initializing Kubernetes configuration
"aws" is not a supported platform for Metrics
Using Profiles service at http://localhost:8084/kfam
Server listening on port http://localhost:8082
Unable to get environment info: Unexpected error getting environment info
Error: connect ECONNREFUSED 127.0.0.1:8084
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)
Unable to get environment info: Unexpected error getting environment info
Error: connect ECONNREFUSED 127.0.0.1:8084
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)
Unable to get environment info: Unexpected error getting environment info
Error: connect ECONNREFUSED 127.0.0.1:8084
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)
Unable to get environment info: Unexpected error getting environment info
Error: connect ECONNREFUSED 127.0.0.1:8084
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)
```

**Environment:**

- Kubeflow version: v0.6.2-rc0
- kfctl version: (use `kfctl version`): v0.6.1-rc.2-1-g3a37cbc6
- Kubernetes platform: (e.g. `minikube`) aws
- Kubernetes version: (use `kubectl version`):  1.13.7
- OS (e.g. from `/etc/os-release`): darwin

**Comment by Jeffwan** [comment]
use `v0.6.0-rc2` instead and then change back to `v20190808-v0.6.2-rc.0-0-g65583158` works. profiles services starts at different port. Not sure why it's not a fixed port.

```
Initializing Kubernetes configuration
"aws" is not a supported platform for Metrics
Using Profiles service at http://10.100.42.90:8081/kfam
Server listening on port http://localhost:8082


```

**Comment by yanniszark** [comment]
Hi @Jeffwan !
The problem is the following:
* Central Dashboard relies on env variables to be populated by Kubernetes, in order to connect to the profiles service.
* The env variables aren't populated when the Central Dashboard Pod starts.
* Central Dashboard uses some default value and continues (localhost:8884).

@avdaredevil how are we going to deal with this?
cc @jlewi

**Comment by jlewi** [comment]
I think this is a bug in the way central dashboard is being implemented.
See comment here: https://github.com/kubeflow/kubeflow/issues/3859#issuecomment-521044568

IIUC the central dashboard is currently relying on Kubernetes to populate environment variables with host and port based on the Kubernetes service. I don't think we need to rely on those environment variables.

We should be able to just rely on Kubernete's internal DNS to route traffic to the KFam backend.
The DNS name for the backend service should be a deterministic function based on the K8s service and/or Virtual service defined for it.

So the K8s application for central dashboard should be able to set it appropriately during `kfctl generate`.

**Comment by jlewi** [comment]
I think this is the problematic code
https://github.com/kubeflow/kubeflow/blob/81fe467f871ec2c08d1b29f50c1a37c5336b902c/components/centraldashboard/app/server.ts#L12

The K8s service for central dashboard is defined here
https://github.com/kubeflow/manifests/blob/master/profiles/base/service.yaml

We should be able to derive from that the K8s DNS name for the KFam service. 
https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#a-records

The only part that I think would vary for different Kubernetes deployments would be the cluster suffix.
It looks like the central dashboard already has a parameter for the suffix
https://github.com/kubeflow/manifests/blob/master/common/centraldashboard/base/params.env


So I think the DNS entry should just

```
kfam.kubeflow.svc.${clusterDomain}
```
@prodonjs Could you prepare a fix to use the DNS entry and not rely on environment variables set by kubernetes?

/assign @prodonjs

**Comment by jlewi** [comment]
@Jeffwan I think you are also picking up changes that should be confined to an RC; i.e. you are picking up the latest centraldashboard image which hasn't been finalized yet.

If you aren't explicitly trying to test an RC; then that's probably a "bug" in that users shouldn't be exposed to upcoming RC's yet.

What was the commands you ran to deploy Kubeflow?

**Comment by jlewi** [comment]
@avdaredevil @prodonjs thoughts?
/assign @avdaredevil

**Comment by prodonjs** [comment]
@jlewi Yes you are correct in spotting where the issue lies. I borrowed the pattern from what we did in setting up the Metadata UI, which itself was derived from the Pipelines UI. I think the difference in those products is that the order of resources is explicitly defined such as in https://github.com/kubeflow/manifests/blob/master/metadata/base/kustomization.yaml, but it still seems like it could be a potential issue depending on when the service actually becomes available.

In any case, I left a [comment](https://github.com/kubeflow/kubeflow/pull/3880#discussion_r313883171) on #3880 which @avdaredevil already had open to address some MUI issues. The suggestion of using the hostname based on the service name should work and because `centraldashboard` and `profiles-kfam` get deployed in the same namespace, the service name is all that's needed to resolve via DNS.

**Comment by Jeffwan** [comment]
> What was the commands you ran to deploy Kubeflow?
@jlewi 
I installed kubeflow on an EKS cluster and all commands are same as I usually run. 

The problem is I didn't override central dashboard version in my config.yaml and it picks up change in 

https://github.com/kubeflow/manifests/blob/f6e898e77c438376d9a3fbec77eb12e819e7b0b6/common/centraldashboard/base/kustomization.yaml#L14-L18

I assume any platform without image version override will have this issue?

**Comment by Jeffwan** [comment]
@prodonjs I think adding initContainer will make the deployment process more reliable,  I've seen this kind of issues frequently for metadata and pipeline projects. I can take the task

/cc @jlewi

**Comment by prodonjs** [comment]
@Jeffwan - I think [Pipelines](https://github.com/kubeflow/pipelines/blob/master/frontend/server/server.ts#L51), [Metadata](https://github.com/kubeflow/metadata/blob/master/frontend/server/server.ts#L29) and centraldashboard could probably all benefit from using the DNS name of the backend services they rely on as opposed to the Kubernetes-provided environment variable which won't be necessarily be set prior to the UI deployment coming up. After exploring the issue, I imagined that had to have happened in other cases so it would be good to get a fix across the board.

**Comment by jlewi** [comment]
@Jeffwan I think the reason you are picking up changes to central dashboard is because you didn't pin to a stable config. The [instructions](https://www.kubeflow.org/docs/aws/deploy/install-kubeflow/) on www.kubeflow.org tell users to use kfctl_aws.yaml

https://raw.githubusercontent.com/kubeflow/kubeflow/master/bootstrap/config/kfctl_aws.yaml 

For stability you should be pointing at the version on the v0.6-branch and possibly at a particular commit.

kfctl_aws.yaml on the v0.6-branch is pinned to a particular commit of kubeflow/manifests
https://github.com/kubeflow/kubeflow/blob/ee47cf1d1d96175697897963b42ebdc9a133a9d8/bootstrap/config/kfctl_gcp_iap.yaml#L299

So using that YAML file would not pick up changes to central dashboard just because we make changes to kubeflow/manifests master or v0.6 branches

**Comment by Jeffwan** [comment]
> @Jeffwan I think the reason you are picking up changes to central dashboard is because you didn't pin to a stable config. The [instructions](https://www.kubeflow.org/docs/aws/deploy/install-kubeflow/) on [www.kubeflow.org](http://www.kubeflow.org) tell users to use kfctl_aws.yaml
> 
> https://raw.githubusercontent.com/kubeflow/kubeflow/master/bootstrap/config/kfctl_aws.yaml
> 
> For stability you should be pointing at the version on the v0.6-branch and possibly at a particular commit.
> 
> kfctl_aws.yaml on the v0.6-branch is pinned to a particular commit of kubeflow/manifests
> https://github.com/kubeflow/kubeflow/blob/ee47cf1d1d96175697897963b42ebdc9a133a9d8/bootstrap/config/kfctl_gcp_iap.yaml#L299
> 
> So using that YAML file would not pick up changes to central dashboard just because we make changes to kubeflow/manifests master or v0.6 branches

Yeah. That's the problem. Instead of controlling version of individual component, I should ping entire registry to a stable version. Thanks! I will have a test and file the PR.

**Comment by jlewi** [comment]
I hit the same issue on GCP.

**Comment by SpontaneousDuck** [comment]
I am experiencing the same issue when deploying on an On-Premise cluster on kubeflow version 0.6.1 and kubernetes version 1.14.3. There are no errors in the log but the namespace list is empty on the dashboard.

**Comment by jlewi** [comment]
@SpontaneousDuck What were the commands you ran to deploy Kubeflow? As noted above this is likely because whatever config you are using isn't pinning to the v0.6.1 release of the dashboard so you are picking up the upcoming RC changes which isn't fully functional yet.

/cc @krishnadurai

**Comment by beatgeek** [comment]
I hit this issue on GCP with centraldashboard on build version v0.6.2-rc.0 of kubeflow and kfctl kfctl v0.6.1-rc.2-1-g3a37cbc6 deploying to an existing cluster.

It sounds like I'm picking up RC changes - I'll pin to 6.1

**Comment by jlewi** [comment]
I think we can close this issue once @Jeffwan submits a change pinning the v0.6.1 AWS config to the appropriate commit of kubeflow/manifests

**Comment by SpontaneousDuck** [comment]
@jlewi I had the problem where it wasn't pinning and changed my config. I verified that the dashboard version was correct after that (v0.6.1) and I still had the same problem.

**Comment by jlewi** [comment]
@SpontaneousDuck What docker image is your centraldashboard using?
Can you provide the logs for the centraldashboard?

Until the DNS issue is fixed you may need to delete the centraldashboard pod to force a restart to get the environment variables

**Comment by jlewi** [comment]
@avdaredevil @prodonjs Has a fix been committed yet to central dashboard to use DNS rather than the environment variables?

**Comment by jlewi** [comment]
it looks like this is fixed by #3899.

@avdaredevil @prodonjs do we need to cherry-pick this onto v0.6-branch and build an image?

**Comment by avdaredevil** [comment]
The build is already made and the PR to merge it is in: https://github.com/kubeflow/manifests/pull/282

**Comment by jlewi** [comment]
Thanks cherry pick is #294

**Comment by jlewi** [comment]
@avdaredevil I'm still seeing this problem with the latest image.

```
gcr.io/kubeflow-images-public/centraldashboard:v20190816-v0.6.0-rc.0-57-g1d4b59c1
```

When I first access central dasbhoard I see an error "can't contact profile controller."

Looking at the central dashboard logs I see

```
> kubeflow-centraldashboard@0.0.2 start /app
> npm run serve


> kubeflow-centraldashboard@0.0.2 serve /app
> node dist/server.js

Initializing Kubernetes configuration
Using Profiles service at http://localhost:8081/kfam
Server listening on port http://localhost:8082
Unable to contact Profile Controller
```

Here's the pod spec
[centraldashboard.pod.yaml.txt](https://github.com/kubeflow/kubeflow/files/3525599/centraldashboard.pod.yaml.txt)

Can we write an E2E test to verify that it is broken and then fixed? For example maybe something like the following

* Add a healthz signal to indicate it is unhealthy or not ready to serve if it can't contact kfam?
* Add a check in https://github.com/kubeflow/kubeflow/blob/master/testing/kfctl/kf_is_ready_test.py to ensure its healthy?

**Comment by prodonjs** [comment]
@avdaredevil - I don't believe NODE_ENV is being set to production in the [Dockerfile](https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/Dockerfile), so the newly added logic that sets the variable for the Profiles service hostname is still reverting to local host.

**Comment by sylus** [comment]
I am still getting this error in latest central dashboard and 0.6.2-rc.1. Is anyone successfully using kubeflow with the latest releases??

**Comment by vanillakola** [comment]
I can see and use other user’s（namespace）servers when I log in by a common user. How to isolate by multi-user. (0.6.2-rc.1)

**Comment by krishnadurai** [comment]
I'm getting a similar error on on-premise server by building kfctl with master branch. I'll verify this with 0.6.2rc1 as well.
'kubeflow-anonymous' namespace gets created now by default - thanks to @kunmingg.

```
Initializing Kubernetes configuration
"other" is not a supported platform for Metrics
Using Profiles service at http://profiles-kfam.kubeflow:8081/kfam
Server listening on port http://localhost:8082
TypeError: name argument is required to req.get
    at IncomingMessage.header (/app/node_modules/express/lib/request.js:67:11)
    at /app/dist/attach_user_middleware.js:11:17
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at SendStream.error (/app/node_modules/serve-static/index.js:121:7)
    at SendStream.emit (events.js:198:13)
    at SendStream.error (/app/node_modules/send/index.js:270:17)
```

**Comment by krishnadurai** [comment]
> I'm getting a similar error on on-premise server by building kfctl with master branch. I'll verify this with 0.6.2rc1 as well.

Facing the same issue with 0.6.2rc1 kfctl_k8s_istio.yaml. The config was the same for master.

**Comment by jlewi** [comment]
@krishnadurai Can you open up a separate issue for the metrics issue? I think the issue there is that the central dashboard can call out to a backend e.g. stackdriver to fetch metrics like cpu that displayed on the home page. I believe if you set it to an appropriate value then it shouldn't try to fetch any issues.

So its likely an issue with properly configuring things for on prem case; e.g. we may need to adjust kfctl_k8s_istio.yaml

**Comment by prodonjs** [comment]
With respect to metrics, there shouldn't be any issue on a non-GKE platform which is currently the only platform that we have a service configured to retrieve metrics, so that message is expected. The other issue looks like it could be do to having empty values for the USERID_HEADER and USERID_PREFIX env variables. Can you get a snapshot of the `env` output from your centraldashboard container?

**Comment by prodonjs** [comment]
@avdaredevil - If these are the default values (https://github.com/kubeflow/manifests/blob/master/common/centraldashboard/base/params.env), then the code in https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/app/attach_user_middleware.ts#L7 will need to be changed to provide default non-empty string values for the header argument. It looks like what's happening is that the call to `req.header()` with an empty string fails.

**Comment by avdaredevil** [comment]
Is this along the lines of what you're talking about @prodonjs 
```js
if (req.headers[userIdHeader]) {
  email = (req.header(userIdHeader)||'').slice(userIdPrefix.length);
  auth = {[userIdHeader]: email};
}
```

**Comment by prodonjs** [comment]
I think you can simplify by just extending the if clause to `if (userIdHeader && req.header(userIdHeader))` and keep the rest of the code the same. Make sure to add a unit test condition as well.

**Comment by carmine** [comment]
This is most likely related, I filed issue #3985 - getting an error in the profiles-deployment pod.

**Comment by jlewi** [comment]
It looks like the latest central dasbhoard image is using DNS.

Here are the logs showing that it is trying to connect to the KFAM service at `http://profiles-kfam.kubeflow:8081/kfam`

```
kubectl -n kubeflow logs centraldashboard-5b8bb76885-gqmcq

> kubeflow-centraldashboard@0.0.2 start /app
> npm run serve


> kubeflow-centraldashboard@0.0.2 serve /app
> node dist/server.js

Initializing Kubernetes configuration
Using Profiles service at http://profiles-kfam.kubeflow:8081/kfam
Server listening on port http://localhost:8082 (in production mode)
jlewi@jlewi-carbon-glaptop:~/git_kubeflow-dev/kubeflow-deployments/jle
```

Docker image

```
image: gcr.io/kubeflow-images-public/centraldashboard:v20190822-v0.6.0-rc.0-68-g127e6006
    imageID: docker-pullable://gcr.io/kubeflow-images-public/centraldashboard@sha256:14a3a23569f4e902368f102478cb1fdda0ea6071445347928aad8f396156271b
```

The container environment variables however look like there was a problem with kustomize substitution.

```
  - env:
    - name: USERID_HEADER
      value: $(userid-header)
    - name: USERID_PREFIX
      value: $(userid-prefix)
    - name: PROFILES_KFAM_SERVICE_HOST
      value: profiles-kfam.kubeflow
```

@avdaredevil  @kunmingg I assume we didn't mean to set USERID_HEADER to ${userid-header)?

My profile is still showing up in the namespace. But it looks like centraldashboard is reporting me as "anonymous@kubeflow.org" (I'm using IAP and logged in as jeremy@lewi.us).

**Comment by avdaredevil** [comment]
Hey Jeremy, not sure if your image includes the Cherrypicked changes, but further debuggability is now available under `/debug` route on the server

**Comment by jlewi** [comment]
I believe this issue has been fixed per my comment above.
It looks like kubeflow/manifests#311 updated the image on the v0.6-branch.
It looks like we are still waiting on kubeflow/manifests#310 to put the change on master.

I think we should wait for kubeflow/manifests#310 to merge to master and then do a quick verification to make sure the issue is fixed and then close this issue out.

**Comment by jlewi** [comment]
Thanks @SpontaneousDuck for confirming; closing this issue.
